janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/util/SSLConfigurationCallback.java

```diff
@@ -0,0 +1,159 @@
+// Copyright 2017 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.es.rest.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.security.KeyManagementException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+
+import javax.net.ssl.SSLContext;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.conn.ssl.NoopHostnameVerifier;
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
+import org.apache.http.ssl.SSLContextBuilder;
+import org.apache.http.ssl.SSLContexts;
+import org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback;
+
+/**
+ * Use {@link Builder#create()} or {@link Builder#createCustom(SSLContextBuilder)} to create an instance of this callback
+ *
+ */
+public class SSLConfigurationCallback implements HttpClientConfigCallback {
+
+    private final String trustStoreFile;
+    private final String trustStorePassword;
+    private final String keyStoreFile;
+    private final String keyStorePassword;
+    private final String keyPassword;
+    private final SSLContextBuilder sslContextBuilder;
+    private final boolean disableHostNameVerification;
+
+    private SSLConfigurationCallback(final SSLContextBuilder sslContextBuilder,
+            final String trustStoreFile,
+            final String trustStorePassword,
+            final String keyStoreFile,
+            final String keyStorePassword,
+            final String keyPassword,
+            final boolean disableHostNameVerification) {
+
+        this.sslContextBuilder = sslContextBuilder;
+
+        this.trustStoreFile = trustStoreFile;
+        this.trustStorePassword = trustStorePassword;
+
+        this.keyStoreFile = keyStoreFile;
+        this.keyStorePassword = keyStorePassword;
+        this.keyPassword = keyPassword;
+
+        this.disableHostNameVerification = disableHostNameVerification;
+    }
+
+    @Override
+    public HttpAsyncClientBuilder customizeHttpClient(HttpAsyncClientBuilder httpClientBuilder) {
+        final SSLContext sslcontext;
+        try {
+            if (StringUtils.isNotEmpty(trustStoreFile)) {
+                sslContextBuilder.loadTrustMaterial(new File(trustStoreFile), trustStorePassword.toCharArray());
+            }
```

Comment:
I still get verification errors with `disableHostNameVerifcation=true`. To resolve I also needed a no-op trust material in this case.

```java
if (StringUtils.isNotEmpty(trustStoreFile) && !disableHostNameVerification) {
    sslContextBuilder.loadTrustMaterial(new File(trustStoreFile), trustStorePassword.toCharArray());
} else if (disableHostNameVerification) {
    sslContextBuilder.loadTrustMaterial(null, (chain,authType) -> true);
}
```

Reply:
Well, then I think we are talking about different features really. Is it possible that you are talking about self-signed certificate support for testing? Because they usually require more than just disabling the host verification. I implemented only the host verification control, which is one of the steps of the SSL strategy.

I have just performed a simple test. I have ES running on a remote host. I use SSH tunnel to connect my JG to it. I have my own custom CA certificate and that server has a certificate for "localhost" signed with this CA - because I intended to connect to it via my localhost (redirected). So, for fun, I have attempted to use an alias to for one of my interfaces instead:

```
Caused by: org.janusgraph.diskstorage.PermanentBackendException: Host name 'mbp-en7' does not match the certificate subject provided by the peer (EMAILADDRESS=ngrigoriev@newforma.com, CN=localhost, OU=Engineering, O=Newforma, L=Montreal, ST=QC, C=CA)
        at org.janusgraph.diskstorage.es.ElasticSearchIndex.<init>(ElasticSearchIndex.java:307)
        ... 31 more
Caused by: javax.net.ssl.SSLPeerUnverifiedException: Host name 'mbp-en7' does not match the certificate subject provided by the peer (EMAILADDRESS=ngrigoriev@newforma.com, CN=localhost, OU=Engineering, O=Newforma, L=Montreal, ST=QC, C=CA)
        at org.apache.http.nio.conn.ssl.SSLIOSessionStrategy.verifySession(SSLIOSessionStrategy.java:208)
        at org.apache.http.nio.conn.ssl.SSLIOSessionStrategy$1.verify(SSLIOSessionStrategy.java:188)
        at org.apache.http.nio.reactor.ssl.SSLIOSession.doHandshake(SSLIOSession.java:368)
        ...
        at java.lang.Thread.run(Thread.java:745)
```

As expected - the CN mismatch. Now, if I enable my newly introduced option, it works perfectly fine without any other changes, it ignores the fact that the certificate presented for "localhost" does not match the host "mbp-en7" I am trying to connect to.

Reply:
Re: PowerMock and coverage - I could not find an easy way to address it. Unfortunately, I had no choice but to use PowerMock for my tests. Some of the classes I needed to deal with, particularly HttpAsyncClientBuilder and RestClientBuilder, are not well designed for testing :( I find it problematic that they do not implement any interfaces, which would allow easy mocking. 

It seems that later versions of PowerMock offer more options - like a  java agent that allows it to coexist with JaCoCo, the current version of PowerMock that does not use Javassist or, worst case, offline instrumentation. But all this also requires newer Mockito. Probably we should create an issue to bump up the versions of the test tools for JG in general.

Reply:
To avoid the arguments, I am implementing another option to enable self-signed certificates.